### PR TITLE
[python.cmake] Detect multiple calls to FINDPYTHON

### PR DESCRIPTION
**Pull request by florent-lamiraux**
Raise FATAL_ERROR is FINDPYTHON is called several times in the same project.
This solves issue https://github.com/jrl-umi3218/jrl-cmakemodules/issues/354.

### CODE REVIEW

**Comment by nim65s** [comment]
This seems a bit excessive. Could we just not do anything, if we detect that this macro has already been called before ?

**Comment by nim65s** [comment]
`FINDPYTHON` is already been called in different places in this module. I think most of them are protected by checking that it has not already been called, but we would have to enforce that.

But I think calling it should not be "FATAL", as this is not by itself a reason to prevent anything to configure or compile. A big annoying AUTHOR_WARNING should do, but we can wait for others to comment :)

**Comment by jcarpent** [comment]
The main question is why two calls of FINDPYTHON are resulting in two different outputs?

**Comment by florent-lamiraux** [comment]
@jcarpent : I do not know.
@nim65s : I tried with ERROR instead of FATAL_ERROR, but I did not find cmake output very expressive. cmake went on configuring and complained that python was not detected correctly.
If we agree on my solution (but I also think we should wait for other people to react), I will add a comment in the documentation of the macro to specify that it should not be called several times.

**Comment by jmirabel** [comment]
FIND_PYTHON may be called several times in metabuild systems like catkin. I agree that we should rather investigate why two calls gives different output. in the meantime, I think we should only add a user warning that the macro was called several times.

@florent-lamiraux were the arguments the same for all the calls ?

**Comment by jmirabel** [comment]
> I think we should only add a user warning that the macro was called several times.

Not a user warning but rather an AUTHOR_WARNING.

**Comment by florent-lamiraux** [comment]
> @florent-lamiraux were the arguments the same for all the calls ?
There was no argument in each call.

**Comment by florent-lamiraux** [comment]
I have just replaced FATAL_ERROR by AUTHOR_WARNING.

File: python.cmake [hunk 1]
```diff
@@ -54,7 +54,10 @@ IF(CMAKE_VERSION VERSION_LESS "3.2")
 ENDIF(CMAKE_VERSION VERSION_LESS "3.2")
 
 MACRO(FINDPYTHON)
-
+  IF(DEFINED FINDPYTHON_ALREADY_CALLED)
+    MESSAGE(AUTHOR_WARNING "Macro FINDPYTHON has already been called.")
```
Comment (jmirabel) [inline python.cmake:58]:
```suggestion
    MESSAGE(AUTHOR_WARNING "Macro FINDPYTHON has already been called. Several call to FINDPYTHON may not find the same Python version (for a yet unknown reason).")
```

**Comment by nim65s** [comment]
Until a better solution is found, this is also fine for me.

But to fix the issue, I suggest to just return from the macro after the AUTHOR_WARNING.

But we can't return from macros in CMake, as they are expanded in place. Could we use a function ? https://cmake.org/cmake/help/latest/command/return.html

**Comment by jmirabel** [comment]
> But to fix the issue, I suggest to just return from the macro after the AUTHOR_WARNING.

This fixes the symptoms, not the issue and it adds one more *IF* to the logic.

**Comment by florent-lamiraux** [comment]
The difference between the first and second calls seems to be in `FIND_PROGRAM(PYTHON_EXECUTABLE python)` in `/usr/share/cmake-3.10/Modules/FindPythonInterp.cmake`.
I do not know how to debug this.

**Comment by florent-lamiraux** [comment]
Fine. I take @jmirabel 's suggestion and merge.
Is it okay for everyone ?

**Pull request merged**